src/staking/RewardTracker.sol

```diff
@@ -0,0 +1,257 @@
+    /// been transferred to this contract before calling this function. If it is called
+    /// when a reward period is still active, a new reward period will begin from the time
+    /// of calling this function, using the leftover rewards from the old reward period plus
+    /// the newly sent rewards as the reward.
+    /// @dev If the reward amount will cause an overflow when computing rewardPerToken, then
+    /// this function will revert.
+    /// @param reward The amount of reward tokens to use in the new reward period.
+    function notifyRewardAmount(uint256 reward) external onlyRole(DISTRIBUTOR) {
+        _notifyRewardAmount(reward);
+    }
+
+    function _notifyRewardAmount(uint256 reward) internal {
+        /// -----------------------------------------------------------------------
+        /// Validation
+        /// -----------------------------------------------------------------------
+
+        if (reward == 0) {
+            return;
+        }
+
+        /// -----------------------------------------------------------------------
+        /// Storage loads
+        /// -----------------------------------------------------------------------
+
+        uint256 rewardRate_ = rewardRate;
+        uint64 periodFinish_ = periodFinish;
+        uint64 lastTimeRewardApplicable_ = block.timestamp < periodFinish_ ? uint64(block.timestamp) : periodFinish_;
+        uint64 DURATION_ = DURATION;
+        uint256 totalSupply_ = totalSupply + totalBonus;
+
+        /// -----------------------------------------------------------------------
+        /// State updates
+        /// -----------------------------------------------------------------------
+
+        // accrue rewards
+        rewardPerTokenStored = _rewardPerToken(totalSupply_, lastTimeRewardApplicable_, rewardRate_);
+        lastUpdateTime = lastTimeRewardApplicable_;
+
+        // record new reward
+        uint256 newRewardRate;
+        if (block.timestamp >= periodFinish_) {
+            newRewardRate = reward / DURATION_;
+        } else {
+            uint256 remaining = periodFinish_ - block.timestamp;
+            uint256 leftover = remaining * rewardRate_;
+            newRewardRate = (reward + leftover) / DURATION_;
+        }
+        // prevent overflow when computing rewardPerToken
+        if (newRewardRate >= ((type(uint256).max / PRECISION) / DURATION_)) {
+            revert Error_AmountTooLarge();
+        }
+        rewardRate = newRewardRate;
+        lastUpdateTime = uint64(block.timestamp);
+        periodFinish = uint64(block.timestamp + DURATION_);
+
+        emit RewardAdded(reward);
+    }
+
+    /// @notice Lets a reward distributor fetch performance fees from
+    /// a vault and start a new reward period.
+    function fetchRewards(ERC4626 vault) external onlyRole(DISTRIBUTOR) {
+        require(isVault[address(vault)], "vault not whitelisted");
```

Comment:
you can introduce a specific error type and use `revert`

Reply:
fixed